### DISCUSSION TITLE
IPv6 support

### DISCUSSION THREAD

**Discussion by thevindu-w** [thevindu-w/clip_share_server#33] [Ideas]
ClipShare currently supports only IPv4. Do we need IPv6 support?

**Comment by Korb** [comment]
@thevindu-w:
> Do you have a use case of ClipShare where IPv4 cannot handle it while IPv6 would do? If so, I'll get back to IPv6 support.

@Korb:
> I have isolation and DHCP server disabled on my MikroTik AP, but it still gives connected devices IP addresses in a different range than the router it is connected to. IPv6 may allow devices to find each other by specifying a specific address, which is not possible now.

@thevindu-w:
> In that case, do the devices get IPv4 adresses? If so, did you try entering the IPv4 address of the server (ex: laptop) in to the address input in the client? Since scanning is not essential if you can enter the address manually. Also, it will save the last used server address and automatically fill the address input with that.
> If the devices get v4 addresses (even though in a different subnet), and if the client can create a TCP connection to the server over IPv4, ClipShare should work in that case. But if the IPv4 addresses are in different subnets and not routable, how does IPv6 do it? Are the IPv6 addresses in the same subnet or somehow routable? If devices don't get IPv4 addresses, then the only solution is IPv6.

Yes, all devices connected directly to the router get IPv4 addresses `192.168.0.0/16`, all devices connected to the AP get IPv4 addresses `10.0.0.0/8`. All devices connected to the router and to the AP get IPv6 addresses. The ISP assigns the prefix 56. I can't say for sure whether this is taken into account when assigning IPv6 addresses by both the router and the AP. At least devices connected to the router and the AP pass all the checks of the online IPv6 test.

An additional nuance is that living in a country with a repressive regime and widespread censorship, I have to constantly use VPN. As I understand it, if the VPN of both devices supports IPv6, using it remains the only way to use ClipShare for devices on different subnets.

Related issue: #32

**Reply by thevindu-w** [reply]
I assume your router and AP are not connected so that there's no route from one network to the other (if there is such route, things would be easier and manageable with IPv4. This is the best case). In the worst case, if you don't have a IPv4 route between those networks, and if you are planning to use IPv6 (after implementing it), still I suggest you not to expose it to the internet (unless it's not in a VPN or other private network or tunnel). Otherwise, requests from the internet can access it making a security threat. In that case, I suggest using one of the following methods
1. Use a VPN or a ssh tunnel or other private network to connect those two networks. In this case, you can get a IPv4 subnet in the VPN eliminating the need of IPv6 (same applies for ssh or other tunnels). The drawback is that if you connect the networks through the internet, there may be latency issues and unreliable connections.
2. If it's possible to connect a device to both networks simultaneously, you can make it to route network traffic across the networks making it possible to use IPv4 while also being in a secure private network. This would be faster than routing through a VPN and internet. However, if your network has restrictions preventing such connections, this option becomes impractical.

**Comment by thevindu-w** [comment]
IPv6 is now supported in the latest versions.
